pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -86,90 +101,370 @@ func (c *ClusterMemberController) sync(ctx context.Context, syncCtx factory.Sync
+	if err != nil {
+		return "", err
+	}
+
+	// Sort nodes by created timestamp.
+	sort.Slice(nodes, func(i, j int) bool {
+		return nodes[i].CreationTimestamp.Before(&nodes[j].CreationTimestamp)
+	})
+
+	// Get the oldest voting member's timestamp
+	var oldestMemberTimestamp *metav1.Time
+	for _, node := range nodes {
+		if _, isMember := memberMap[node.Name]; isMember {
+			oldestMemberTimestamp = &node.CreationTimestamp
+		}
+	}
+	if oldestMemberTimestamp.IsZero() {
+		return "", fmt.Errorf("could not find any node mapped to a member")
+	}
+
+	// Only scale up if current membership is less than the max allowed
+	currentMemberCount := len(members)
+	maxMemberCount := c.replicaCount + 1 // Surge up
+	if currentMemberCount >= maxMemberCount {
+		klog.Warningf("Skipping scale up etcd membership at maximum size: %d", maxMemberCount)
+		return "", nil
+	}
+
+	for _, node := range nodes {
+		// Ignore if the node is mapped to a member and already part of the quorum.
+		_, isMember := memberMap[node.Name]
+		if isMember {
+			klog.V(4).Infof("node/%s is already mapped to an etcd member", node.Name)
 			continue
 		}
-		isEtcdContainerRunning, isEtcdContainerReady := false, false
-		for _, containerStatus := range pod.Status.ContainerStatuses {
-			if containerStatus.Name != "etcd" {
+
+		// To avoid re-adding a recently scaled-down/member-less node back into the cluster
+		// filter out any nodes older than the current oldest member node.
+		// We assume those older nodes to already be scaled-down.
+		if node.CreationTimestamp.Before(oldestMemberTimestamp) {
+			// TODO(haseeb): What happens if the node is the same age as the oldest member node?
+			// Could be a recently removed node or it could be the next new node we need to scale up (e.g 3 new nodes created at the same time)
+			// In practice the nodes seem to differ in their CreationTimestamps but there should be a better way to mark scaled-down nodes.
+			//
+			// Is it safe to assume that the to-be-scaled-down nodes are always deleted first?
+			// If so, the Machine deletion timestamp could be used.
+			//
+			// Or a new Node/Machine annotation or condition added by the etcd-operator on scaled-down nodes.
+			continue
+		}
+
+		internalIP, err := dnshelpers.GetEscapedPreferredInternalIPAddressForNodeName(network, node)
+		if err != nil {
+			return "", fmt.Errorf("failed to get internal IP for node: %w", err)
+		}
+		peerURL := fmt.Sprintf("https://%s:2380", internalIP)
+		return peerURL, nil
+	}
+
+	return "", nil
+}
+
+func (c *ClusterMemberController) ensureEtcdLearnerPromotion(ctx context.Context, recorder events.Recorder) error {
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return fmt.Errorf("could not get etcd member list: %v", err)
+	}
+
+	var errs []error
+	for _, member := range members {
+		// Promote learner to voting member if log is in sync with leader.
+		if member.IsLearner {
```

Comment:
Can we get a list of the Machine to member mapping here and check the presence of the deletion hook before we promote? This should prevent us ever promoting a member that doesn't have the hook if there's a bug elsewhere in the logic